File: packages/app/src/lib/server/core/dotenv.ts
import { SecretKey } from "@prick/shared";

import { PrickError } from "./errors.js";

/**
 * A strict `.env` parser.
 *
 * "Strict" is the entire specification. Every `.env` parser in the wild is
 * lenient in a different direction, and leniency in a secrets manager is not a
 * convenience -- it is a class of outage. The three decisions that follow from
 * that:
 *
 * 1. **NO `$VAR` INTERPOLATION. AT ALL.**
 *    Several popular parsers expand `$OTHER` / `${OTHER}` inside double quotes.
 *    Here they do not, and there is no flag to make them. Consider what silent
 *    interpolation means for this product: a value of `$aB3!x$k` is a perfectly
 *    ordinary generated password, and an interpolating parser stores it as
 *    `$aB3!x` plus whatever `$k` happened to resolve to -- usually the empty
 *    string. The secret is now WRONG, it is wrong in a way that still looks like
 *    a password, and nothing anywhere reports an error. A parser that stores
 *    what the file says cannot do that.
 *
 *    A `${...}` sequence still produces a WARNING, because someone migrating
 *    from an interpolating tool needs to be told their file will not mean what
 *    it used to. The warning names the LINE and never the text.
 *
 * 2. **Every failure carries a line number.** "Invalid .env file" for a 300-line
 *    paste is not a diagnosis.
 *
 * 3. **A duplicate key is an ERROR, not last-one-wins.** A file that sets
 *    `DATABASE_URL` twice is a file whose author has already lost track of it;
 *    silently picking one is picking which of two production databases the
 *    deploy talks to.
 *
 * CRLF is tolerated everywhere, including inside a quoted multi-line value,
 * where it is NORMALISED to `\n`. A secret's bytes must not depend on whether
 * the file came off a Windows checkout.
 *
 * Multi-line values are possible ONLY inside quotes. A bare newline ends an
 * unquoted value, always.
 */

export interface DotenvEntry {
  key: string;
  value: string;
  /** 1-based line the key was declared on. */
  line: number;
}

export interface DotenvWarning {
  line: number;
  key: string;
  message: string;
}

export interface DotenvDocument {
  entries: DotenvEntry[];
  /** Non-fatal observations. The UI shows these on the import dry-run screen. */
  warnings: DotenvWarning[];
}

/**
 * A parse failure, carrying the line it happened on.
 *
 * `VALIDATION_FAILED` (422) rather than `BAD_REQUEST`: the request was
 * well-formed HTTP carrying a well-formed body; it is the CONTENT of a field
 * that is unacceptable, which is exactly what 422 means.
 *
 * THE MESSAGE NEVER QUOTES THE LINE. It says what is wrong and where, and stops
 * -- the offending line of a `.env` file is, by construction, a line containing
 * a secret, and this message travels into an HTTP response and a Worker log.
 */
export class DotenvParseError extends PrickError {
  readonly line: number;

  constructor(line: number, message: string, hint?: string) {
    super("VALIDATION_FAILED", `Line ${String(line)}: ${message}`, {
      ...(hint === undefined ? {} : { hint }),
      detail: { line },
    });
    this.name = "DotenvParseError";
    this.line = line;
  }
}

/** Recognised escapes inside a double-quoted value. */
const DOUBLE_QUOTE_ESCAPES: Record<string, string> = {
  n: "\n",
  r: "\r",
  t: "\t",
  f: "\f",
  b: "\b",
  '"': '"',
  "'": "'",
  "\\": "\\",
  $: "$",
  "`": "`",
};

interface Cursor {
  readonly text: string;
  index: number;
  line: number;
}

export function parseDotenv(source: string): DotenvDocument {
  // A UTF-8 BOM in front of the first key turns `KEY` into `﻿KEY`, which
  // then fails the POSIX name check with a message about an invalid character
  // that is invisible in every editor.
  const text = source.charCodeAt(0) === 0xfeff ? source.slice(1) : source;

  const cursor: Cursor = { text, index: 0, line: 1 };
  const entries: DotenvEntry[] = [];
  const warnings: DotenvWarning[] = [];
  const seen = new Map<string, number>();

  while (cursor.index < cursor.text.length) {
    skipBlank(cursor);
    if (cursor.index >= cursor.text.length) break;

    const declaredOn = cursor.line;
    const key = readKey(cursor);
    const value = readValue(cursor, key);

    const previous = seen.get(key);
    if (previous !== undefined) {
      throw new DotenvParseError(
        declaredOn,
        `duplicate key "${key}", already set on line ${String(previous)}.`,
        "Remove one of the two declarations. This file does not say which value you meant.",
      );
    }
    seen.set(key, declaredOn);

    if (/\$\{[^}]*\}|\$[A-Za-z_][A-Za-z0-9_]*/.test(value)) {
      warnings.push({
        line: declaredOn,
        key,
        // Names the KEY (plaintext metadata) and never any part of the value.
        message:
          "value contains a $VAR-like sequence and will be stored literally -- this parser performs no interpolation.",
      });
    }

    entries.push({ key, value, line: declaredOn });
  }

  return { entries, warnings };
}

/** Advance past whitespace, blank lines and whole-line comments. */
function skipBlank(cursor: Cursor): void {
  while (cursor.index < cursor.text.length) {
    const char = cursor.text[cursor.index];

    if (char === "\n") {
      cursor.index += 1;
      cursor.line += 1;
      continue;
    }

    if (char === "\r" || char === " " || char === "\t") {
      cursor.index += 1;
      continue;
    }

    if (char === "#") {
      skipToEndOfLine(cursor);
      continue;
    }

    return;
  }
}

function skipToEndOfLine(cursor: Cursor): void {
  while (cursor.index < cursor.text.length && cursor.text[cursor.index] !== "\n") {
    cursor.index += 1;
  }
}

/**
 * Read `KEY` (optionally behind `export `) up to and including the `=`.
 *
 * `export ` is accepted because half the `.env` files in existence were written
 * to be `source`d by a shell. It is accepted and then DISCARDED -- it carries no
 * meaning here.
 */
function readKey(cursor: Cursor): string {
  const startLine = cursor.line;
  const start = cursor.index;

  skipToEndOfLine(cursor);
  const rawLine = cursor.text.slice(start, cursor.index).replace(/\r$/, "");

  const equals = rawLine.indexOf("=");
  if (equals === -1) {
    throw new DotenvParseError(
      startLine,
      "expected KEY=value but found no `=`.",
      "Every non-comment line must assign a value. Multi-line values are only possible inside quotes.",
    );
  }

  let name = rawLine.slice(0, equals).trim();
  if (name.startsWith("export ") || name.startsWith("export\t")) {
    name = name.slice("export".length).trim();
  }

  if (name === "") {
    throw new DotenvParseError(startLine, "the name before `=` is empty.");
  }

  const validated = SecretKey.safeParse(name);
  if (!validated.success) {
    // Names the KEY. A key name is plaintext metadata everywhere in this system
    // -- it is what the UI lists and the audit log records -- so echoing it is
    // not a leak, and without it the operator has to count lines to find out
    // which one this is about.
    throw new DotenvParseError(
      startLine,
      `"${name}" is not a valid environment variable name.`,
      "Names must be a letter or underscore followed by letters, digits or underscores.",
    );
  }

  // Rewind to just after the `=` so the value reader starts there. The line
  // counter is untouched: `skipToEndOfLine` never crossed a newline.
  cursor.index = start + equals + 1;

  return validated.data;
}

function readValue(cursor: Cursor, key: string): string {
  // Leading horizontal whitespace between `=` and the value is insignificant,
  // so the opening quote is looked for PAST it -- but the cursor is only
  // advanced for a quoted value. `readUnquoted` needs to see that whitespace,
  // because whether it was there is exactly what distinguishes
  // `COLOR=#ffffff` (a value) from `COLOR= # a comment` (an empty one).
  let peek = cursor.index;
  while (cursor.text[peek] === " " || cursor.text[peek] === "\t") peek += 1;

  const opener = cursor.text[peek];

  if (opener === '"') {
    cursor.index = peek;
    return readDoubleQuoted(cursor, key);
  }
  if (opener === "'") {
    cursor.index = peek;
    return readSingleQuoted(cursor, key);
  }
  if (opener === "`") {
    cursor.index = peek;
    throw new DotenvParseError(
      cursor.line,
      `the value of "${key}" is backtick-quoted, which is not accepted.`,
      "Use single quotes for a literal value or double quotes if you need escapes.",
    );
  }

  return readUnquoted(cursor);
}

/**
 * An unquoted value: everything to end of line, minus a trailing comment.
 *
 * THE COMMENT RULE, stated exactly because it is the ambiguous case every
 * parser gets differently: a `#` starts a comment ONLY when preceded by
 * whitespace. So `COLOR=#ffffff` is the seven-character value `#ffffff`;
 * `COLOR=#ffffff # the brand one` is that same value with a comment; and
 * `COLOR= # nothing yet` is the EMPTY value with a comment. A `#` immediately
 * following a non-space character is part of the value, which is what keeps
 * `TOKEN=ab#cd` from silently becoming `ab`.
 *
 * This is why `readValue` hands the leading whitespace over intact rather than
 * consuming it: the search is for whitespace-then-`#`, and consuming the
 * whitespace first would make the empty-value case indistinguishable from the
 * `#ffffff` one.
 *
 * If you need a value that both starts with whitespace and contains ` #`, quote
 * it. That is what quotes are for.
 */
function readUnquoted(cursor: Cursor): string {
  const start = cursor.index;
  skipToEndOfLine(cursor);

  const raw = cursor.text.slice(start, cursor.index).replace(/\r$/, "");

  const comment = raw.search(/[ \t]#/);
  const body = comment === -1 ? raw : raw.slice(0, comment);

  return body.trim();
}

function readSingleQuoted(cursor: Cursor, key: string): string {
  const openedOn = cursor.line;
  cursor.index += 1;

  let out = "";

  while (cursor.index < cursor.text.length) {
    const char = cursor.text[cursor.index];

    if (char === "'") {
      cursor.index += 1;
      assertNothingAfterClose(cursor, key);
      return out;
    }

    if (char === "\r") {
      // CRLF normalisation: a value must not change because of a checkout's
      // line-ending policy.
      if (cursor.text[cursor.index + 1] === "\n") {
        cursor.index += 1;
        continue;
      }
    }

    if (char === "\n") cursor.line += 1;

    out += char;
    cursor.index += 1;
  }

  throw new DotenvParseError(
    openedOn,
    `the single-quoted value of "${key}" is never closed.`,
    "A single-quoted value is literal -- there are no escapes, so it cannot contain a single quote.",
  );
}

function readDoubleQuoted(cursor: Cursor, key: string): string {
  const openedOn = cursor.line;
  cursor.index += 1;

  let out = "";

  while (cursor.index < cursor.text.length) {
    const char = cursor.text[cursor.index];

    if (char === "\\") {
      const escaped = cursor.text[cursor.index + 1];

      if (escaped === undefined) break;

      if (escaped === "\n") {
        // A backslash-newline is a line continuation and contributes nothing.
        cursor.index += 2;
        cursor.line += 1;
        continue;
      }

      const replacement = DOUBLE_QUOTE_ESCAPES[escaped];
      if (replacement === undefined) {
        throw new DotenvParseError(
          cursor.line,
          `unknown escape sequence in the value of "${key}".`,
          "Recognised escapes are \\n \\r \\t \\f \\b \\\" \\' \\\\ \\$ and \\`. A lone backslash must be written \\\\.",
        );
      }

      out += replacement;
      cursor.index += 2;
      continue;
    }

    if (char === '"') {
      cursor.index += 1;
      assertNothingAfterClose(cursor, key);
      return out;
    }

    if (char === "\r" && cursor.text[cursor.index + 1] === "\n") {
      cursor.index += 1;
      continue;
    }

    if (char === "\n") cursor.line += 1;

    out += char;
    cursor.index += 1;
  }

  throw new DotenvParseError(
    openedOn,
    `the double-quoted value of "${key}" is never closed.`,
    "Check for an unescaped double quote inside the value.",
  );
}

/**
 * After a closing quote only whitespace or a comment may follow.
 *
 * `KEY="a" garbage` is rejected rather than being read as `a`. Accepting it
 * means a mistyped `KEY="a" "b"` silently stores `a`, and the operator has no
 * way to see that half their value was discarded.
 */
function assertNothingAfterClose(cursor: Cursor, key: string): void {
  while (cursor.index < cursor.text.length) {
    const char = cursor.text[cursor.index];

    if (char === " " || char === "\t" || char === "\r") {
      cursor.index += 1;
      continue;
    }

    if (char === "\n" || char === "#") return;

    throw new DotenvParseError(
      cursor.line,
      `unexpected text after the closing quote of "${key}".`,
      "Only whitespace or a `#` comment may follow a quoted value.",
    );
  }
}

/**
 * Parse into the map shape the write path takes.
 *
 * Separate from `parseDotenv` so the import dry-run can report warnings and line
 * numbers, which a bare `Record<string, string>` cannot carry.
 */
export function parseDotenvToMap(source: string): Record<string, string> {
  const out: Record<string, string> = {};
  for (const entry of parseDotenv(source).entries) out[entry.key] = entry.value;
  return out;
}

File: crates/prick-core/src/dotenv.rs
//! A strict `.env` parser.
//!
//! `.env` has no specification, so every implementation disagrees at the edges.
//! This one resolves the ambiguity in the direction a secrets manager needs:
//! **a line that cannot be parsed unambiguously is an error, never a guess.**
//! Importing a file and silently dropping the two lines the parser did not
//! understand is how a deploy loses `DATABASE_URL`.
//!
//! The accepted grammar:
//!
//! | Form | Meaning |
//! |---|---|
//! | `# ...` or blank | Ignored |
//! | `KEY=value` | Unquoted; surrounding whitespace trimmed, no escapes |
//! | `KEY='value'` | Literal; no escape sequences at all, `\` is a backslash |
//! | `KEY="value"` | `\\ \" \n \r \t` are escapes, everything else is literal |
//! | `export KEY=...` | The `export ` prefix is accepted and dropped |
//!
//! A quoted value may be followed by `# comment`, because the closing quote has
//! already said where the value ends.
//!
//! On an **unquoted** value, whitespace before a `#` is the one line shape this
//! format cannot resolve, and it is [rejected](DotenvError::AmbiguousComment).
//! `PASSWORD=hunter2 # 1` is either a password containing a hash or a password
//! with a comment after it; both occur in real files, and either reading stores
//! a value the author did not write without saying so. The two quoted forms
//! state which was meant, and `PASSWORD=hunter2#1` -- no whitespace -- is
//! unambiguous already.
//!
//! # Where this grammar is spoken
//!
//! `prk secrets upload` hands the document to the server, which parses it there
//! and reports the line it refused; that parser accepts this grammar and the
//! same rule about `#`. What this one does locally is prove the other
//! direction: [`crate::format::render`] emits `.env` output, and its tests read
//! that output back through here, so a downloaded environment is one this
//! product can be handed again.

use crate::keyname;

/// Why a `.env` document could not be parsed.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[non_exhaustive]
pub enum DotenvError {
    /// A non-blank, non-comment line contained no `=`.
    #[error("line {line}: expected `KEY=value`")]
    MissingEquals {
        /// One-based line number.
        line: usize,
    },
    /// The key on this line was not a legal key name.
    #[error("line {line}: {source}")]
    InvalidKey {
        /// One-based line number.
        line: usize,
        /// The underlying validation failure.
        source: keyname::KeyNameError,
    },
    /// A quoted value opened but never closed.
    #[error("line {line}: unterminated {quote} quote")]
    UnterminatedQuote {
        /// One-based line number.
        line: usize,
        /// Which quote character was left open.
        quote: char,
    },
    /// A double-quoted value contained an escape this parser does not define.
    #[error("line {line}: unknown escape `\\{ch}` (valid: \\\\ \\\" \\n \\r \\t)")]
    UnknownEscape {
        /// One-based line number.
        line: usize,
        /// The character that followed the backslash.
        ch: char,
    },
    /// A quoted value had trailing content after the closing quote.
    #[error("line {line}: unexpected text after the closing quote")]
    TrailingContent {
        /// One-based line number.
        line: usize,
    },
    /// An unquoted value had whitespace before a `#`, which is a comment or
    /// part of the value with nothing in the line to say which.
    #[error("line {line}: unquoted value has a `#` after whitespace; quote the value to keep it")]
    AmbiguousComment {
        /// One-based line number.
        line: usize,
    },
    /// The same key appeared twice. Last-one-wins is a silent data loss bug.
    #[error("line {line}: duplicate key `{key}`, first defined on line {first}")]
    DuplicateKey {
        /// One-based line number of the second definition.
        line: usize,
        /// One-based line number of the first definition.
        first: usize,
        /// The repeated key.
        key: String,
    },
}

/// Parses a `.env` document into key/value pairs, preserving file order.
///
/// # Errors
///
/// Any line that is not unambiguously parseable aborts the whole parse. A
/// partially imported `.env` is worse than a rejected one.
pub fn parse(input: &str) -> Result<Vec<(String, String)>, DotenvError> {
    let mut out: Vec<(String, String)> = Vec::new();
    let mut seen: Vec<(&str, usize)> = Vec::new();

    for (index, raw) in input.lines().enumerate() {
        let line = index + 1;
        let trimmed = raw.trim();

        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }

        let trimmed = trimmed.strip_prefix("export ").map_or(trimmed, str::trim_start);

        let (key, rest) = trimmed.split_once('=').ok_or(DotenvError::MissingEquals { line })?;
        let key = key.trim_end();

        keyname::validate(key).map_err(|source| DotenvError::InvalidKey { line, source })?;

        if let Some((_, first)) = seen.iter().find(|(k, _)| *k == key) {
            return Err(DotenvError::DuplicateKey { line, first: *first, key: key.to_owned() });
        }
        seen.push((key, line));

        let value = parse_value(rest, line)?;
        out.push((key.to_owned(), value));
    }

    Ok(out)
}

/// Parses the right-hand side of a `KEY=` assignment.
///
/// The opening quote is looked for *past* any whitespace, but [`parse_unquoted`]
/// is handed that whitespace intact: whether it was there is what separates
/// `COLOR=#ffffff`, a value, from `COLOR= #ffffff`, which could be that same
/// value written with insignificant whitespace or an empty one with a comment.
fn parse_value(rest: &str, line: usize) -> Result<String, DotenvError> {
    let trimmed = rest.trim_start();
    match trimmed.as_bytes().first() {
        Some(b'\'') => parse_single_quoted(&trimmed[1..], line),
        Some(b'"') => parse_double_quoted(&trimmed[1..], line),
        _ => parse_unquoted(rest, line),
    }
}

/// An unquoted value runs to the end of the line, minus surrounding whitespace.
///
/// Whitespace before a `#` makes the line unreadable either way round, so it is
/// refused rather than resolved. A `#` that follows a non-space character is
/// part of the value: `TOKEN=ab#cd` is `ab#cd`.
fn parse_unquoted(rest: &str, line: usize) -> Result<String, DotenvError> {
    if rest.contains(" #") || rest.contains("\t#") {
        return Err(DotenvError::AmbiguousComment { line });
    }
    Ok(rest.trim().to_owned())
}

/// Single quotes are fully literal: the value ends at the next `'`.
fn parse_single_quoted(body: &str, line: usize) -> Result<String, DotenvError> {
    let end = body.find('\'').ok_or(DotenvError::UnterminatedQuote { line, quote: '\'' })?;
    check_after_close(&body[end + 1..], line)?;
    Ok(body[..end].to_owned())
}

/// After a closing quote only whitespace or a `#` comment may follow.
///
/// `KEY="a" b` is refused rather than read as `a`, so a mistyped `KEY="a" "b"`
/// cannot store half a value without saying so. A comment is fine: the quotes
/// already delimited the value, which is what makes quoting the answer to
/// [`DotenvError::AmbiguousComment`].
fn check_after_close(rest: &str, line: usize) -> Result<(), DotenvError> {
    let rest = rest.trim_start();
    if rest.is_empty() || rest.starts_with('#') {
        Ok(())
    } else {
        Err(DotenvError::TrailingContent { line })
    }
}

/// Double quotes support exactly five escapes and reject any other.
fn parse_double_quoted(body: &str, line: usize) -> Result<String, DotenvError> {
    let mut out = String::with_capacity(body.len());
    let mut chars = body.chars();

    while let Some(ch) = chars.next() {
        match ch {
            '"' => {
                check_after_close(chars.as_str(), line)?;
                return Ok(out);
            }
            '\\' => {
                let escaped =
                    chars.next().ok_or(DotenvError::UnterminatedQuote { line, quote: '"' })?;
                out.push(match escaped {
                    '\\' => '\\',
                    '"' => '"',
                    'n' => '\n',
                    'r' => '\r',
                    't' => '\t',
                    other => return Err(DotenvError::UnknownEscape { line, ch: other }),
                });
            }
            other => out.push(other),
        }
    }

    Err(DotenvError::UnterminatedQuote { line, quote: '"' })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parsed(input: &str) -> Vec<(String, String)> {
        parse(input).unwrap()
    }

    #[test]
    fn parses_a_plain_document() {
        let out = parsed("A=1\nB=two\n");
        assert_eq!(out, [("A".to_owned(), "1".to_owned()), ("B".to_owned(), "two".to_owned())]);
    }

    #[test]
    fn preserves_file_order() {
        let out = parsed("Z=1\nA=2\nM=3\n");
        let keys: Vec<&str> = out.iter().map(|(k, _)| k.as_str()).collect();
        assert_eq!(keys, ["Z", "A", "M"]);
    }

    #[test]
    fn skips_comments_and_blank_lines() {
        let out = parsed("# leading\n\n  \nA=1\n# trailing\n");
        assert_eq!(out.len(), 1);
    }

    #[test]
    fn accepts_the_export_prefix() {
        assert_eq!(parsed("export A=1\n"), [("A".to_owned(), "1".to_owned())]);
        // Not a prefix without the space.
        assert!(parse("exportA=1\n").is_ok());
        assert_eq!(parsed("exportA=1\n")[0].0, "exportA");
    }

    #[test]
    fn a_hash_after_whitespace_in_an_unquoted_value_is_refused_not_guessed_at() {
        // Both readings of this line are losses. Truncating it stores "hunter2"
        // and the deploy authenticates with the wrong password; keeping it whole
        // stores a value with a stray comment welded onto the end. Neither is
        // visible to the operator, so the line is refused instead.
        assert_eq!(parse("PASSWORD=hunter2 # 1\n"), Err(DotenvError::AmbiguousComment { line: 1 }));
        assert_eq!(parse("A=1\nB=2 \t# x\n"), Err(DotenvError::AmbiguousComment { line: 2 }));
        // An empty value with a comment reads the same two ways: the value could
        // be nothing, or `#ffffff` written with insignificant whitespace.
        assert_eq!(parse("COLOR= # ffffff\n"), Err(DotenvError::AmbiguousComment { line: 1 }));
    }

    #[test]
    fn a_hash_that_follows_a_non_space_stays_in_the_value() {
        // The load-bearing case for the rule above: a parser that strips from
        // the first `#` stores the empty string for a colour and half a token.
        assert_eq!(parsed("COLOR=#ffffff\n")[0].1, "#ffffff");
        assert_eq!(parsed("TOKEN=ab#cd\n")[0].1, "ab#cd");
    }

    #[test]
    fn quoting_is_how_a_value_keeps_its_hash() {
        // The way out of `AmbiguousComment`, in both directions.
        assert_eq!(parsed("PASSWORD=\"hunter2 # 1\"\n")[0].1, "hunter2 # 1");
        assert_eq!(parsed("PASSWORD='hunter2 # 1'\n")[0].1, "hunter2 # 1");
        assert_eq!(parsed("PASSWORD=\"hunter2\" # 1\n")[0].1, "hunter2");
    }

    #[test]
    fn unquoted_values_are_trimmed() {
        assert_eq!(parsed("A=1   \n")[0].1, "1");
        assert_eq!(parsed("A=   1\n")[0].1, "1");
        assert_eq!(parsed("A=\n")[0].1, "");
        assert_eq!(parsed("A=   \n")[0].1, "");
    }

    #[test]
    fn single_quotes_are_entirely_literal() {
        assert_eq!(parsed(r"A='a\nb'")[0].1, r"a\nb");
        assert_eq!(parsed(r#"A='he said "hi"'"#)[0].1, r#"he said "hi""#);
        assert_eq!(parsed("A='  spaced  '")[0].1, "  spaced  ");
        assert_eq!(parsed("A=''")[0].1, "");
    }

    #[test]
    fn double_quotes_interpret_the_five_escapes() {
        assert_eq!(parsed(r#"A="a\nb""#)[0].1, "a\nb");
        assert_eq!(parsed(r#"A="a\tb""#)[0].1, "a\tb");
        assert_eq!(parsed(r#"A="a\rb""#)[0].1, "a\rb");
        assert_eq!(parsed(r#"A="a\\b""#)[0].1, r"a\b");
        assert_eq!(parsed(r#"A="a\"b""#)[0].1, r#"a"b"#);
    }

    #[test]
    fn double_quotes_preserve_everything_else_literally() {
        assert_eq!(parsed(r#"A="$HOME `id` ${x}""#)[0].1, "$HOME `id` ${x}");
        assert_eq!(parsed(r#"A="  spaced  ""#)[0].1, "  spaced  ");
    }

    #[test]
    fn unknown_escapes_are_errors_not_silent_passthrough() {
        assert_eq!(parse(r#"A="a\qb""#), Err(DotenvError::UnknownEscape { line: 1, ch: 'q' }));
        // A unicode escape is an unknown escape too. The `.env` ecosystem
        // treats `\u` inconsistently, so accepting one here would produce a
        // value other tools read back differently. The backslash is
        // interpolated so the escape never appears literally in this source
        // file.
        let unicode_escape = format!("A=\"a{}u0041b\"", '\\');
        assert_eq!(parse(&unicode_escape), Err(DotenvError::UnknownEscape { line: 1, ch: 'u' }));
    }

    #[test]
    fn unterminated_quotes_are_errors() {
        assert_eq!(parse("A='oops"), Err(DotenvError::UnterminatedQuote { line: 1, quote: '\'' }));
        assert_eq!(
            parse(r#"A="oops"#),
            Err(DotenvError::UnterminatedQuote { line: 1, quote: '"' })
        );
        assert_eq!(
            parse(r#"A="oops\"#),
            Err(DotenvError::UnterminatedQuote { line: 1, quote: '"' })
        );
    }

    #[test]
    fn trailing_content_after_a_closing_quote_is_an_error() {
        assert_eq!(parse(r#"A="a" b"#), Err(DotenvError::TrailingContent { line: 1 }));
        assert_eq!(parse("A='a' b"), Err(DotenvError::TrailingContent { line: 1 }));
        // Trailing whitespace is fine.
        assert_eq!(parsed(r#"A="a"   "#)[0].1, "a");
    }

    #[test]
    fn a_comment_may_follow_a_closing_quote() {
        // The closing quote already said where the value ended, so there is
        // nothing left to guess at -- and this is what makes "quote it" a real
        // answer for someone whose file has comments on its secret lines.
        assert_eq!(parsed(r#"A="a" # why"#)[0].1, "a");
        assert_eq!(parsed("A='a'   # why")[0].1, "a");
        assert_eq!(parsed(r##"A="a"#comment"##)[0].1, "a");
    }

    #[test]
    fn missing_equals_is_an_error() {
        assert_eq!(parse("A=1\nJUST_A_WORD\n"), Err(DotenvError::MissingEquals { line: 2 }));
    }

    #[test]
    fn invalid_keys_are_rejected_with_a_line_number() {
        assert!(matches!(parse("1BAD=x"), Err(DotenvError::InvalidKey { line: 1, .. })));
        assert!(matches!(parse("A=1\nb-c=x"), Err(DotenvError::InvalidKey { line: 2, .. })));
    }

    #[test]
    fn duplicate_keys_are_rejected_rather_than_last_one_wins() {
        assert_eq!(
            parse("A=1\nB=2\nA=3\n"),
            Err(DotenvError::DuplicateKey { line: 3, first: 1, key: "A".to_owned() })
        );
    }

    #[test]
    fn key_may_have_whitespace_before_the_equals() {
        assert_eq!(parsed("A = 1\n"), [("A".to_owned(), "1".to_owned())]);
    }

    #[test]
    fn crlf_input_does_not_leak_carriage_returns_into_values() {
        // str::lines strips \r\n, so a CRLF checkout of a fixture parses the
        // same as an LF one. Worth pinning: an invisible \r on the end of a
        // token produces baffling authentication failures.
        assert_eq!(
            parsed("A=1\r\nB=2\r\n"),
            [("A".to_owned(), "1".to_owned()), ("B".to_owned(), "2".to_owned())]
        );
    }
}
